# Goal of the ETL Lesson

The main goal of this ETL pipelines lesson is to take the [World Bank Project data set](https://datacatalog.worldbank.org/dataset/world-bank-projects-operations) and merge this data with the [World Bank indicator data](https://data.worldbank.org/indicator/SP.POP.TOTL). Then you'll load the merged data into a database.

In the process, you'll need to transform these data sets in different ways. And finally, you'll code an ETL pipeline to extract, transform, and load the data all in one step.

# Extracting data from a csv file

The first step in an ETL pipeline is extraction. Data comes in all types of different formats, and you'll practice extracting data from csv files, JSON files, XML files, SQL databases, and the web.

In this first exercise, you'll practice extracting data from a CSV file and then navigating through the results. You'll see that extracting data is not always a straight-forward process.

This exercise contains a series of coding questions for you to solve. If you get stuck, there is a solution file called 1_csv_exercise_solution.ipynb. You can find this solution file by going to File->Open and then clicking on the file name.

# Part 1 projects_data.csv

You'll be using the following csv files:
* projects_data.csv
* population_data.csv

As a first step, try importing the projects data using the pandas [read_csv method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html). The file path is just '../data/projects_data.csv'. You can see the file if you click on File->Open in the workspace and open the data folder.

In [1]:
# TODO: import the projects_data.csv file using the pandas library
# Store the results in the df_projects variable
import pandas as pd
df_projects = pd.read_csv("projects_data.csv")

/opt/conda/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (44) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [2]:
df_projects.head()

id  regionname                                        countryname  \
0  P162228       Other                                        World;World   
1  P163962      Africa  Democratic Republic of the Congo;Democratic Re...   
2  P167672  South Asia  People's Republic of Bangladesh;People's Repub...   
3  P158768  South Asia  Islamic  Republic of Afghanistan;Islamic  Repu...   
4  P161364      Africa  Federal Republic of Nigeria;Federal Republic o...   

  prodline                  lendinginstr lendinginstrtype  \
0       RE  Investment Project Financing               IN   
1       PE  Investment Project Financing               IN   
2       PE  Investment Project Financing               IN   
3       PE  Investment Project Financing               IN   
4       PE  Investment Project Financing               IN   

  envassesmentcategorycode supplementprojectflg productlinetype  \
0                        C                    N               L   
1                        B                    N               L   
2                      NaN                    Y               L   
3                        A                    N               L   
4                        B                    N               L   

  projectstatusdisplay     ...     mjtheme3name mjtheme4name mjtheme5name  \
0               Active     ...              NaN          NaN          NaN   
1               Active     ...              NaN          NaN          NaN   
2               Active     ...              NaN          NaN          NaN   
3               Active     ...              NaN          NaN          NaN   
4               Active     ...              NaN          NaN          NaN   

                                            location  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4  0002327546!$!Ogun State!$!7!$!3.58333!$!NG;000...   

                                            GeoLocID  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4  0002327546;0002328925;0002565340;0002565343;00...   

                                          GeoLocName                 Latitude  \
0                                                NaN                      NaN   
1                                                NaN                      NaN   
2                                                NaN                      NaN   
3                                                NaN                      NaN   
4  Ogun State;Niger State;Abia State;Edo;Kebbi St...  7;10;5.41667;6.5;11.5;8   

                Longitude            Country Unnamed: 56  
0                     NaN                NaN         NaN  
1                     NaN                NaN         NaN  
2                     NaN                NaN         NaN  
3                     NaN                NaN         NaN  
4  3.58333;6;7.5;6;4;10.5  NG;NG;NG;NG;NG;NG         NaN  

[5 rows x 57 columns]

Did you get a DType warning? Read about what this warning is in the [pandas documentation](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.errors.DtypeWarning.html).

Pandas tries to figure out programatically the data type of each column (integer, float, boolean, string). In this case, pandas could not automatically figure out the data type. That is because some columns have more than one possible data types. In other words, this data is messy.

You can use the dtype option to specify the data type of each column. Because there are so many columns in this data set, you can set all columns to be strings at least for now.

Try reading in the data set again using the read_csv() method. This time, also use the option dtype=str so that pandas treats everything like a string.

In [3]:
# TODO: Read in the projects_data.csv file using the read_csv method 
# and dtype = str option
df_projects = pd.read_csv("projects_data.csv", dtype=str)

In [4]:
# Run the cell below to see what the data looks like
df_projects.head()

id  regionname                                        countryname  \
0  P162228       Other                                        World;World   
1  P163962      Africa  Democratic Republic of the Congo;Democratic Re...   
2  P167672  South Asia  People's Republic of Bangladesh;People's Repub...   
3  P158768  South Asia  Islamic  Republic of Afghanistan;Islamic  Repu...   
4  P161364      Africa  Federal Republic of Nigeria;Federal Republic o...   

  prodline                  lendinginstr lendinginstrtype  \
0       RE  Investment Project Financing               IN   
1       PE  Investment Project Financing               IN   
2       PE  Investment Project Financing               IN   
3       PE  Investment Project Financing               IN   
4       PE  Investment Project Financing               IN   

  envassesmentcategorycode supplementprojectflg productlinetype  \
0                        C                    N               L   
1                        B                    N               L   
2                      NaN                    Y               L   
3                        A                    N               L   
4                        B                    N               L   

  projectstatusdisplay     ...     mjtheme3name mjtheme4name mjtheme5name  \
0               Active     ...              NaN          NaN          NaN   
1               Active     ...              NaN          NaN          NaN   
2               Active     ...              NaN          NaN          NaN   
3               Active     ...              NaN          NaN          NaN   
4               Active     ...              NaN          NaN          NaN   

                                            location  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4  0002327546!$!Ogun State!$!7!$!3.58333!$!NG;000...   

                                            GeoLocID  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4  0002327546;0002328925;0002565340;0002565343;00...   

                                          GeoLocName                 Latitude  \
0                                                NaN                      NaN   
1                                                NaN                      NaN   
2                                                NaN                      NaN   
3                                                NaN                      NaN   
4  Ogun State;Niger State;Abia State;Edo;Kebbi St...  7;10;5.41667;6.5;11.5;8   

                Longitude            Country Unnamed: 56  
0                     NaN                NaN         NaN  
1                     NaN                NaN         NaN  
2                     NaN                NaN         NaN  
3                     NaN                NaN         NaN  
4  3.58333;6;7.5;6;4;10.5  NG;NG;NG;NG;NG;NG         NaN  

[5 rows x 57 columns]

In [5]:
# TODO: count the number of null values in the data set
# HINT: use the isnull() and sum() methods
df_projects.isnull().sum()

id                              0
regionname                      0
countryname                     0
prodline                        0
lendinginstr                  246
lendinginstrtype              246
envassesmentcategorycode     5811
supplementprojectflg           53
productlinetype                 0
projectstatusdisplay            4
status                          4
project_name                    0
boardapprovaldate            1504
board_approval_month         1504
closingdate                  3349
lendprojectcost               125
ibrdcommamt                     0
idacommamt                      0
totalamt                        0
grantamt                        0
borrower                     5919
impagency                    6097
url                             0
projectdoc                  18248
majorsector_percent         18248
sector1                         0
sector2                      8721
sector3                     11761
sector4                     13872
sector5       

Notice that the number 18248 shows up multiple times. There is also a countryname column with 0 missing values and a Country column with 14045 missing values. This data set clearly has some issues that will need to be solved in the transform part of the pipeline.

Next, output the shape of the data frame.

In [6]:
# TODO: output the shape of the data frame
df_projects.shape

(18248, 57)

There are 18248 rows in this data set. Considering many columns had 18248 NaN values, many columns in the data set are filled completely with NaN values.

# Part 2 population_data.csv

Next, use the pandas read_csv method to read in the population_data.csv file. The path to this file is "../data/population_data.csv". When you try to read in this data set using pandas, you'll get an error because there is something wrong with the data.

In [7]:
# TODO: read in the population_data.csv file using the read_csv() method
# Put the results in a variable called df_population
df_population = pd.read_csv("population_data.csv")

ParserError: Error tokenizing data. C error: Expected 3 fields in line 5, saw 63


There is something wrong with this data set. You should see an error that says "expected 3 fields in line 5, saw 63". What might have happened? Try printing out the first few lines of the data file to see what the issue might be.

In [ ]:
# TODO: Print out the first 10 lines of the data set, line by line.
with open("population_data.csv", 'r') as f:
    for _ in range(10):
        print("line", _, ":", f.readline())

The first four lines in the file are not properly formatted and don't contain data. Next, read in the data using the read_csv method. But this time, use the skiprows option.

In [ ]:
# TODO: read in population data skipping the first four rows
# Put the results in a variable called df_population

df_population = pd.read_csv("population_data.csv", skiprows=4)

In [ ]:
# Run this cell to see what the data looks like
df_population.head()

Make sure to scroll over to see what the last column looks like. That last column, called 'Unnamed: 62', doesn't look very useful and is filled with NaN values.

In [ ]:
# TODO: Count the number of null values in each column
df_population.isnull().sum()

It looks like every year column has at least one NaN value. 

In [ ]:
# TODO: Count the number of null values in each row
# HINT: In the sum method, use axis=1
df_population.isnull().sum(axis=1)

And it looks like almost every row has only one null value. That is probably from the 'Unnamed: 62' column that doesn't have any relevant information in it. Next, drop the 'Unnamed: 62' column from the data frame.

In [ ]:
# TODO: drop the 'Unnamed: 62' column from the data frame, 
# and save the results in the df_population variable

df_population = df_population.drop(['Unnamed: 62'], axis=1)

In [ ]:
# Run this code cell.  
# This code outputs any row that contains a null value
# The purpose is to see what rows contain null values now that 
#   'Unnamed: 62' was dropped from the data.
df_population[df_population.isnull().any(axis=1)]

# Conclusion 

This population data doesn't look too bad. Only six rows have missing values. In the transformation part of the lesson, you'll have to deal with these missing values somehow.

If you would like to see the solution file for this exercise, go to File->Open and click on 1_csv_exercise_solution.ipynb.

In the next exercise, you'll practice extracting data json and xml files.